tensorflow/core/kernels/non_max_suppression_op.cu.cc

```diff
@@ -0,0 +1,433 @@
+                                          &host_nms_mask, alloc_attr));
+
+    // this will return sorted scores and their indices
+    config = GetCudaLaunchConfig(num_boxes, device);
+    // initialize box and score indices
+    Iota<<<config.block_count, config.thread_per_block, 0, device.stream()>>>(
+        config, 0, d_indices.flat<int>().data());
+    cuda_ret = cudaGetLastError();
+    CUDA_CHECK(cuda_ret);
+    CHECK_EQ(cuda_ret, cudaSuccess);
+    cuda_ret = cub::DeviceRadixSort::SortPairsDescending(
+        d_cub_sort_buffer.flat<int8>().data(), cub_sort_temp_storage_bytes,
+        scores.flat<float>().data(), d_sorted_scores.flat<float>().data(),
+        d_indices.flat<int>().data(), d_sorted_indices.flat<int>().data(),
+        num_boxes, 0,
+        8 * sizeof(float),  // sort all bits
+        cuda_stream);
+    CUDA_CHECK(cuda_ret);
+    CHECK_EQ(cuda_ret, cudaSuccess);
+
+    // get pointers for easy access
+    const float4* original_boxes =
+        reinterpret_cast<const float4*>(boxes.flat<float>().data());
+    float4* sorted_boxes =
+        reinterpret_cast<float4*>(d_sorted_boxes.flat<float>().data());
+    const int* sorted_indices = d_sorted_indices.flat<int>().data();
+    // sort boxes using indices
+    IndexSelect<<<config.block_count, config.thread_per_block, 0,
+                  device.stream()>>>(config, original_boxes, sorted_indices,
+                                     sorted_boxes);
+    CHECK_EQ(cuda_ret, cudaSuccess);
+    int num_to_keep = 0;
+    bool flip_boxes = true;  // there is no guarantee that boxes are given in
+                             // lower left-upper right corners!
+    auto status =
+        nms_gpu(d_sorted_boxes.flat<float>().data(), num_boxes,
+                iou_threshold_val, d_selected_indices.flat<int>().data(),
+                &num_to_keep, d_nms_mask.flat<int>().data(),
+                host_nms_mask.flat<int>().data(), context, flip_boxes);
+    cuda_ret = cudaGetLastError();
+    CUDA_CHECK(cuda_ret);
+    CHECK_EQ(cuda_ret, cudaSuccess);
+    if (!status.ok()) {
+      context->SetStatus(status);
+      return;
+    }
+    Tensor* output_indices = nullptr;
+    int num_outputs = std::min(num_to_keep, output_size);  // no padding!
+    OP_REQUIRES_OK(context,
+                   context->allocate_output(0, TensorShape({num_outputs}),
+                                            &output_indices));
+    if (num_outputs == 0) return;
+    config = GetCudaLaunchConfig(num_outputs, device);
+    IndexSelect<<<config.block_count, config.thread_per_block, 0,
+                  device.stream()>>>(config, sorted_indices,
+                                     d_selected_indices.flat<int>().data(),
+                                     (*output_indices).flat<int>().data());
+    CHECK_EQ(cudaGetLastError(), cudaSuccess);
+  }
+};
+REGISTER_KERNEL_BUILDER(
+    Name("NonMaxSuppressionV2").TypeConstraint<float>("T").Device(DEVICE_GPU),
```

Comment:
That will take time. I will add v3 and v4 in subsequent PRs. Regardless, NMSKernel and NmsGpu are required for #28754 and need to stay as they are. V2 is the closest user of current kernels thus easier to add.

Reply:
@aaroey Also all these https://github.com/tensorflow/models/search?p=1&q=non_max_suppression&unscoped_q=non_max_suppression can be switched to use v2 op on GPU with a simple regex replace (and should IMHO).

Reply:
They actually all use v3 ops: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_impl.py#L2582. So I would say this implementation probably won't be useful if it's for v2 only.

Reply:
But isn't V3 with score_threshold=-inf and pad_to_max_output_size=False == V2? I don't see score_threshold and pad_to_max_output being passed around in most places. Just adding _v2 to most of these should give a significant perf boost. Only the one in post_processing.py is making use of V3, the rest can interchangeably use V2 or V3 since V3 with default last 2 arguments is equivalent to v2.

Reply:
As we discussed in person, I am keeping this and will implement other ops in a later PR.